### PR TITLE
Upgrade to React Native 0.72

### PR DESCRIPTION
**Pull request by krystofwoldrich**
- Supersedes https://github.com/sentry-demos/sentry_react_native/pull/34
- Supersedes https://github.com/sentry-demos/sentry_react_native/pull/7
- Closes https://github.com/sentry-demos/sentry_react_native/issues/36
- Closes https://github.com/sentry-demos/sentry_react_native/issues/37
- Closes https://github.com/sentry-demos/sentry_react_native/issues/39
- Closes https://github.com/sentry-demos/sentry_react_native/issues/38

### This PR does
- Upgrade to 0.72 RN including bump of all related dependencies to up to date version compatible with the latest RN
- Upgrade to Sentry SDK v5
- Fix all lint issues
- Fix some TS types issues (`npx tsc --noEmit` rest of current issues will be fixed in a future PR)
- Fix use of SAGP
- Remove iOS artifacts from the source code
- Enables screenshots and view hierarchy
- Node requirement to 20.x
- Ruby requirement to 3.2.2
- Npm requirement to 9.8.0

### This PR does not 
- Add any Sentry feature
- Fix any demo features
- Fix all TS types issues
- Update Readme and Install guides

### Warning
- This might break the automated tests (wd removed from deps - was not used in the app code, native components structure might changed)

### Testing
- [Error before](https://github.com/sentry-demos/sentry_react_native/pull/35#issuecomment-1700552975)
- [Error after](https://github.com/sentry-demos/sentry_react_native/pull/35#issuecomment-1700560202)
- [Transaction before](https://github.com/sentry-demos/sentry_react_native/pull/35#issuecomment-1700585575)
- [Transaction after](https://github.com/sentry-demos/sentry_react_native/pull/35#issuecomment-1700588230)

### After this PR
- Bump the app version to 3.0.0

### CODE REVIEW

File: android/settings.gradle [hunk 1]
```diff
@@ -1,5 +1,4 @@
 rootProject.name = 'sentry_react_native'
-include ':react-native-linear-gradient'
-project(':react-native-linear-gradient').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-linear-gradient/android')
```
Comment (krystofwoldrich) [inline android/settings.gradle:3]: The latest version of `react-native-linear-gradient` doesn't mention that this is needed.

File: babel.config.js [hunk 2]
```diff
@@ -1,13 +1,4 @@
 module.exports = {
   presets: ['module:metro-react-native-babel-preset'],
-  plugins: [
-    ['module:react-native-dotenv', {
-      'moduleName': '@env',
-      'path': '.env',
-      'blacklist': null,
-      'whitelist': null,
-      'safe': false,
-      'allowUndefined': true
```
Comment (krystofwoldrich) [inline babel.config.js:10]: These are the defaults. So they don't have to be included.

File: index.js [hunk 3]
```diff
@@ -2,6 +2,7 @@
  * @format
  */
 
+import 'react-native-gesture-handler';
```
Comment (krystofwoldrich) [inline index.js:5]: Specified by the lib that it should be the first import in the project.

File: android/app/src/main/java/com/sentry_react_native/MainActivity.java [hunk 4]
```diff
@@ -12,4 +17,23 @@ public class MainActivity extends ReactActivity {
...
+        this,
+        getMainComponentName(),
+        // If you opted-in for the New Architecture, we enable the Fabric Renderer.
+        DefaultNewArchitectureEntryPoint.getFabricEnabled());
+  }
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(null);
+  }
```
Comment (krystofwoldrich) [inline android/app/src/main/java/com/sentry_react_native/MainActivity.java:38]: Required by react-native-screens

**Comment by thinkocapo** [comment]
I'm curious the before-vs-after of this PR. What did the issue details page show before this PR, and what does it show after? Did we gain anything? Did we lose anything?

Same for the transaction titled 'Home'

**Comment by krystofwoldrich** [comment]
[Example error before this PR](https://testorg-az.sentry.io/issues/3568862458/events/4d82e8fcd5794217969d7b779a010f9e/)
screencapture-testorg-az-sentry-io-issues-3568862458-events-latest-2023-08-31-09_54_05

**Comment by krystofwoldrich** [comment]
[Example error after this PR](https://testorg-az.sentry.io/issues/4438316682/events/76547b7b3c564b49937fc24111991ba9/)

- Correct stack trace
- More App information
- React Native Context information
- Packages list
- Screenshot is present
- View Hierarchy is present

screencapture-testorg-az-sentry-io-issues-4438316682-events-latest-2023-08-31-09_52_01

**Comment by krystofwoldrich** [comment]
[Home transaction before this PR](https://testorg-az.sentry.io/performance/mobile-react-native:7b9628d9112a4c659a6f00ff7862e735/)

screencapture-testorg-az-sentry-io-performance-mobile-react-native-7b9628d9112a4c659a6f00ff7862e735-2023-08-31-10_22_22

**Comment by krystofwoldrich** [comment]
[Home Transaction after this PR](https://testorg-az.sentry.io/performance/mobile-react-native:08884adb093d47ff92e4701db1c23b30/)

screencapture-testorg-az-sentry-io-performance-mobile-react-native-08884adb093d47ff92e4701db1c23b30-2023-08-31-10_25_06

File: src/App.tsx [hunk 5]
```diff
@@ -24,75 +25,81 @@ import Toast from 'react-native-toast-message';
...
-  // debug: true
+  attachStacktrace: true,
+  attachScreenshot: true,
+  attachViewHierarchy: true,
 });
 
 Sentry.setTag('se', SE);
 
-const Stack = createStackNavigator();
+const Stack = createStackNavigator<RootStackParamList>();
```
Comment (thinkocapo) [inline src/App.tsx:80]: is this for an upgraded version of ReactNavigation?

**Review by thinkocapo — APPROVED** [review]

File: src/App.tsx (hunk 5 above)
Comment (krystofwoldrich) [reply src/App.tsx:80]: I believe it was in the previous version too. It just improves the typings of the props in the screen components. Like [here](https://github.com/sentry-demos/sentry_react_native/pull/35/files#diff-aafd64a88abcae3bdb91b28ad88a0eacd48be83e285bbf6b0c0eb3cfe29d96b2R59-R61). So the `navigator` is not implicitly `any`.

**Pull request merged**